src/Text/Pandoc/Writers/Typst.hs

```diff
@@ -88,6 +88,52 @@ pandocToTypst options (Pandoc meta blocks) = do
        Nothing  -> main
        Just tpl -> renderTemplate tpl context
 
+pickTypstAttrs :: [(Text, Text)] -> ([(Text, Text)],[(Text, Text)])
+pickTypstAttrs =
+  partition (not . T.isPrefixOf "typst:text:" . fst) . filter (T.isPrefixOf "typst:" . fst)
+
+formatTypstProps :: [(Text, Text)] -> [Text]
+formatTypstProps =
+  map (\(k,v) -> last (T.splitOn (T.pack ":") k) <> ": " <> v)
+
+toTypstProps :: [(Text, Text)] -> Doc Text
+toTypstProps typstAttrs =
+  literal (T.intercalate ", " $ formatTypstProps typstAttrs)
+
+toTypstPropsListUnchecked :: [(Text, Text)] -> Doc Text
+toTypstPropsListUnchecked typstAttrs =
+  toTypstProps typstAttrs <> ","
+
+toTypstTextElementUnchecked :: [(Text, Text)] -> Doc Text -> Doc Text
+toTypstTextElementUnchecked typstTextAttrs content =
+  "#text" <> parens (toTypstProps typstTextAttrs) <> brackets content
+
+toTypstSetTextUnchecked :: [(Text, Text)] -> Doc Text
+toTypstSetTextUnchecked typstTextAttrs =
+  "#set text" <> parens (toTypstProps typstTextAttrs) <> "; " -- newline?
+
+-- there is probably a more idiomatic way to do this
+ifNotEmpty ::  ([a2] -> Doc Text) -> [a2] -> Doc Text
+ifNotEmpty f list =
+  case list of
+    [] -> ""
+    _ -> f list
+
```

Comment:
I think it would be more elegant to use a fold.  E.g. `foldl' addTypstProperty mempty props`.
If the list `props` is empty, you get `mempty`, so you wouldn't need this ugly function.
It should be possible to make this code simpler.

Reply:
It's a little bit more complicated, because we have this conditional text we want to add around the lists, like we want to emit either `#set text(prop:value); ` or the empty string at the top of a block where we are applying `typst:text:*` attributes. We don't want to emit a useless `#set text(); ` if that's even valid.

And likewise everywhere else: if a block has no attributes, we want to emit `#block[content]` and not `#block()[content]`

@cscheid and I looked at it together and agreed that most clear thing to do is not to try to abstract it, but just write functions that both check for the empty list and generate text. It's about half as long now.